cuda_core/cuda/core/experimental/_dlpack.pyx

```diff
@@ -91,14 +76,103 @@ cpdef object make_py_capsule(object buf, bint versioned):
+        for i in range(ndim):
+            shape_strides[i] = layout.base.shape[i]
+        if layout.base.strides == NULL:
+            dl_tensor.strides = NULL
+        else:
+            dl_tensor.strides = shape_strides + ndim
+            for i in range(ndim):
+                dl_tensor.strides[i] = layout.base.strides[i]
+        dl_tensor.byte_offset = 0
+    return 0
+
+
+cdef inline int _setup_dl_tensor_dtype(DLTensor* dl_tensor, object dtype) except -1:
+    cdef DLDataType* dl_dtype = &dl_tensor.dtype
+    if dtype is None:
+        dl_dtype.code = <uint8_t>kDLInt
+        dl_dtype.lanes = <uint16_t>1
+        dl_dtype.bits = <uint8_t>8
+        return 0
+    cdef uint8_t code
+    cdef uint8_t bits
+    cdef uint16_t lanes
+    code, bits, lanes = dtype
+    dl_dtype.code = code
+    dl_dtype.bits = bits
+    dl_dtype.lanes = lanes
+    return 0
+
+
+cpdef object make_py_capsule(
+    object buf,
+    bint versioned,
+    intptr_t data_ptr,
+    StridedLayout layout=None,
+    object dtype=None,
+):
+    cdef DLManagedTensor* dlm_tensor = NULL
+    cdef DLManagedTensorVersioned* dlm_tensor_ver = NULL
+    cdef DLTensor* dl_tensor = NULL
+    cdef void* tensor_ptr
+    cdef const char* capsule_name
 
-    cdef DLDataType* dtype = &dl_tensor.dtype
-    dtype.code = <uint8_t>kDLInt
-    dtype.lanes = <uint16_t>1
-    dtype.bits = <uint8_t>8
-
-    cpython.Py_INCREF(buf)
-    return cpython.PyCapsule_New(tensor_ptr, capsule_name, pycapsule_deleter)
+    try:
+        if versioned:
+            dlm_tensor_ver = <DLManagedTensorVersioned*>(
+                stdlib.malloc(sizeof(DLManagedTensorVersioned)))
+            dlm_tensor_ver.dl_tensor.shape = NULL
+            dl_tensor = &dlm_tensor_ver.dl_tensor
+            dlm_tensor_ver.version.major = DLPACK_MAJOR_VERSION
+            dlm_tensor_ver.version.minor = DLPACK_MINOR_VERSION
+            cpython.Py_INCREF(buf)
+            dlm_tensor_ver.manager_ctx = <void*>buf
+            dlm_tensor_ver.deleter = versioned_deleter
+            dlm_tensor_ver.flags = 0
+            tensor_ptr = dlm_tensor_ver
+            capsule_name = DLPACK_VERSIONED_TENSOR_UNUSED_NAME
+        else:
+            dlm_tensor = <DLManagedTensor*>(
+                stdlib.malloc(sizeof(DLManagedTensor)))
+            dlm_tensor.dl_tensor.shape = NULL
+            dl_tensor = &dlm_tensor.dl_tensor
+            cpython.Py_INCREF(buf)
+            dlm_tensor.manager_ctx = <void*>buf
+            dlm_tensor.deleter = deleter
+            tensor_ptr = dlm_tensor
+            capsule_name = DLPACK_TENSOR_UNUSED_NAME
+
+        dl_tensor.data = <void*>data_ptr
+
+        _setup_dl_tensor_device(dl_tensor, buf)
+        _setup_dl_tensor_layout(dl_tensor, buf, layout)
+        _setup_dl_tensor_dtype(dl_tensor, dtype)
+
+        return cpython.PyCapsule_New(tensor_ptr, capsule_name, pycapsule_deleter)
+    except:
+        cleanup(dlm_tensor)
+        cleanup_versioned(dlm_tensor_ver)
```

Comment:
those two are almost like the deleters, just with extra ifs to handle incomplete initialization in case of exceptions.